tests/EdgeToEdge/TrackBannerDonationRedirectsTest.php

```diff
@@ -0,0 +1,202 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Tests\EdgeToEdge;
+
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use WMDE\Fundraising\DonationContext\Domain\Model\Donation;
+use WMDE\Fundraising\DonationContext\Tests\Data\ValidDonation;
+use WMDE\Fundraising\Frontend\App\EventHandlers\TrackBannerDonationRedirects;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\FixedTokenGenerator;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\SessionSpy;
+
+/**
+ * @covers \WMDE\Fundraising\Frontend\App\EventHandlers\TrackBannerDonationRedirects
+ */
+class TrackBannerDonationRedirectsTest extends WebRouteTestCase {
+
+	private const CORRECT_ACCESS_TOKEN = 'KindlyAllowMeAccess';
+	private const ADD_ROUTE = 'add_donation';
+	private const CONFIRMATION_ROUTE = 'show_donation_confirmation';
+	private const DISALLOWED_ROUTE = 'show-donation-form';
+	private const URL_PARAMETER = 'banner_submission';
+	private const TEST_CAMPAIGN = 'test_campaign';
+	private const TEST_KEYWORD = 'test_keyword';
+
+	private SessionInterface $session;
+	private KernelBrowser $client;
+	private Donation $donation;
+
+	public function setUp(): void {
+		$this->session = new SessionSpy();
```

Comment:
This could spare you the lengthy re-implementation of the session interface:

```suggestion
		$this->session = new Session(new MockArraySessionStorage());
```